src/components/data/data-utils.js

```diff
@@ -472,68 +472,217 @@ export const EntityRenamer = ({ entityType, entityName, workspaceId: { namespace
-    [initialIsReference, () => attributeValue.entityType],
-    () => entityTypes[0]
-  ))
-  const [editType, setEditType] = useState(() => Utils.cond(
-    [initialIsReference, () => 'reference'],
-    // explicit double-equal to check for null and undefined, since entity attribute lists can contain nulls
-    // eslint-disable-next-line eqeqeq
-    [(initialIsList ? attributeValue.items[0] : attributeValue) == undefined, () => 'string'],
-    [initialIsList, () => typeof attributeValue.items[0]],
-    () => typeof attributeValue
-  ))
-  const [isBusy, setIsBusy] = useState()
-  const [consideringDelete, setConsideringDelete] = useState()
+  return { type, isList }
+}
 
-  const isList = _.isArray(newValue)
-  const isUnchanged = (initialType === editType && initialIsList === isList && _.isEqual(attributeValue, newValue))
+export const convertAttributeValue = (attributeValue, newType, referenceEntityType) => {
+  if (newType === 'reference' && !referenceEntityType) {
+    throw new Error('An entity type is required to convert an attribute to a reference')
+  }
 
-  const makeTextInput = placeholder => ({ value = '', ...props }) => h(TextInput, { autoFocus: true, placeholder, value, ...props })
+  const { type, isList } = getAttributeType(attributeValue)
 
-  const { prepForUpload, makeInput, blankVal } = Utils.switchCase(editType,
-    ['string', () => ({
-      prepForUpload: _.trim,
-      makeInput: makeTextInput('Enter a value'),
-      blankVal: ''
-    })],
-    ['reference', () => ({
-      prepForUpload: v => ({ entityName: _.trim(v), entityType: linkedEntityType }),
-      makeInput: makeTextInput(`Enter a ${linkedEntityType}_id`),
-      blankVal: ''
-    })],
-    ['number', () => ({
-      prepForUpload: _.identity,
-      makeInput: ({ value = 0, ...props }) => h(NumberInput, { autoFocus: true, isClearable: false, value, ...props }),
-      blankVal: 0
+  const baseConvertFn = Utils.switchCase(
+    newType,
+    ['reference', () => value => ({
+      entityType: referenceEntityType,
+      entityName: _.toString(value) // eslint-disable-line lodash-fp/preferred-alias
     })],
-    ['boolean', () => ({
-      prepForUpload: _.identity,
-      makeInput: ({ value = false, ...props }) => div({ style: { flexGrow: 1, display: 'flex', alignItems: 'center', height: '2.25rem' } },
-        [h(Switch, { checked: value, ...props })]),
-      blankVal: false
-    })]
+    ['number', () => value => {
+      const numberVal = _.toNumber(value)
+      return _.isNaN(numberVal) ? 0 : numberVal
+    }],
+    [Utils.DEFAULT, () => Utils.convertValue(newType)]
   )
+  const convertFn = type === 'reference' ? _.flow(_.get('entityName'), baseConvertFn) : baseConvertFn
+
+  if (isList) {
+    return _.flow(
+      _.update('items', _.map(convertFn)),
+      newType === 'reference' ? _.set('itemsType', 'EntityReference') : _.omit('itemsType')
```

Comment:
Lists of entity references look like:
```
{
   items: [
     { entityType: 'thing', entityName: 'thing_1' },
     { entityType: 'thing', entityName: 'thing_1' }
  ],
  itemsType: 'EntityReference'
}
```

This was dropping that `itemsType` field when converting from a list of references to a list of some other type. However, I missed that lists of other types also have an `itemsType` field, but it's set to `AttributeValue` instead of `EntityReference`.

This conversion becomes more clear after [e8870fd](https://github.com/DataBiosphere/terra-ui/pull/2895/commits/e8870fdb45a008185b03a88c1cb6db2bfd6ad678).